src/components/NotesMindMap/NotesMindMapReducer.js

```diff
@@ -14,114 +14,128 @@ import {
 import { ROOT_NOTE_FOUND_ACTION } from 'components/App/AppActions.js';
 import {
   updateGroupOfOldParent,
-  addGroupTagToNote,
-  removeNoteFromGraph,
+  addGroupTagToNodes,
+  removeNodeFromGraph,
 } from '../../helpers/graph';
 
 const defaultState = {
   selectedNote: {},
   showNoteNameEditor: false,
   noteText: '',
   isChangeParentModeActive: false,
-  notes: [],
+  nodes: [],
   edges: [],
 };
 
 export const notesMindMapReducer = (state = defaultState, { type, data }) => {
   const handleSelectedNoteChildrenFetchedAction = () => {
-    let notes = [...state.notes];
+    let nodes = [...state.nodes];
     let edges = [...state.edges];
 
     const childNotes = data;
     if (childNotes.length) {
       childNotes.forEach(child => {
-        notes.push({
-          id: child.id,
-          label: child.name,
-          name: child.name,
-          isNote: child.isNote,
-          group: 'children',
-        });
+        nodes = addNoteToGraph(nodes, child);
         edges.push({ from: child.parent.id, to: child.id });
       });
-      notes = addGroupTagToNote(notes, state.selectedNote.id);
+      nodes = addGroupTagToNodes(nodes, edges);
     }
-    return { ...state, notes, edges };
+    return {
+      ...state,
+      nodes,
+      edges,
+    };
   };
 
   const handleEditNoteNameAction = () => {
     const noteToEdit = data;
-    return { ...state, showNoteNameEditor: true, selectedNote: noteToEdit };
+    return {
+      ...state,
+      showNoteNameEditor: true,
+      selectedNote: noteToEdit,
+    };
   };
 
   const handleNoteNameUpdateRequestSuccessAction = () => {
     const updatedNote = data;
-    let notes = [...state.notes];
-    notes = notes.filter(note => note.id !== updatedNote.id);
-    notes.push({
-      id: updatedNote.id,
-      label: updatedNote.name,
-      name: updatedNote.name,
-      isNote: updatedNote.isNote,
-    });
-    return { ...state, notes };
+    let nodes = [...state.nodes];
+    nodes = nodes.filter(note => note.id !== updatedNote.id);
+    return {
+      ...state,
+      nodes: addNoteToGraph(nodes, updatedNote),
+    };
   };
 
   const handleCreateNoteSuccessAction = () => {
     const newNote = data;
-    let notes = [...state.notes];
+    let nodes = [...state.nodes];
     let edges = [...state.edges];
-    notes.push({
-      id: newNote.id,
-      name: newNote.name,
-      label: newNote.name,
-      isNote: newNote.isNote,
-      group: 'children',
-    });
     edges.push({ from: newNote.parent.id, to: newNote.id });
-    notes = addGroupTagToNote(notes, state.selectedNote.id);
-    return { ...state, notes, edges };
+    nodes = addGroupTagToNodes(addNoteToGraph(nodes, newNote), edges);
```

Comment:
When we create a note, its parent should either keep the "parent" group, or change the group from "children" to "parent", if it's the first child of the node.

So we need to check only 1 node: the parent node. We don't need to map all the nodes.

Reply:
Done